### PR TITLE
[SPARK-24478][SQL][followup] Move projection and filter push down to physical conversion

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/21503, to completely move operator pushdown to the planner rule.

The code are mostly from https://github.com/apache/spark/pull/21319

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rdblue @gengliangwang

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 1]
```diff
@@ -105,117 +105,57 @@ case class StreamingDataSourceV2Relation(
...
-      source match {
-        case support: ReadSupportWithSchema =>
-          support
-        case _: ReadSupport =>
-          throw new AnalysisException(
-            s"Data source does not support user-supplied schema: $name")
-        case _ =>
-          throw new AnalysisException(s"Data source is not readable: $name")
-      }
+    private def asReadSupport: ReadSupport = source match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: strip one indentation level to shorten the code.

**Comment by cloud-fan** [comment]
retest this please

**Comment by rxin** [comment]
Does this move actually make sense? It'd destroy stats estimation for partition pruning.

**Comment by cloud-fan** [comment]
According to the [dicsussion](https://github.com/apache/spark/pull/21503#discussion_r194871032) , we agree that it's better to remove the hacky workaround for mutable `DataSourceV2Relation` and not having stats estimation for partition pruning. For builtin data source like file source, we can still keep the optimzer rule `PruneFileSourcePartitions` to make them work.

**Comment by rdblue** [comment]
@rxin, we can also add the second pushdown (in the stats visitor) to get better stats with a property to turn it on or off. We're going to add it back in our branch anyway.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: I'm -1 on cosmetic changes like this. It doesn't make a big improvement, but makes it harder to track down when a functional change happened.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala [hunk 2]
```diff
@@ -17,51 +17,115 @@
...
-        (projectSet ++ filterSet).toSeq
-      }
 
-      val reader = relation.newReader
+  /**
+   * Pushes down filters to the data source reader
+   *
+   * @return pushed filter and post-scan filters.
+   */
+  private def pushFilters(
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala]: +1 for moving these functions. I considered it in the other commit, but decided to go with fewer changes. I like them here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 3]
```diff
@@ -23,17 +23,24 @@ import org.apache.spark.sql.AnalysisException
...
+ * @param options The options for this scan. Used to create fresh [[DataSourceReader]].
+ * @param userSpecifiedSchema The user-specified schema for this scan. Used to create fresh
+ *                            [[DataSourceReader]].
+ */
 case class DataSourceV2Relation(
     source: DataSourceV2,
     output: Seq[AttributeReference],
     options: Map[String, String],
-    userSpecifiedSchema: Option[StructType] = None)
+    userSpecifiedSchema: Option[StructType])
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: Why is this change necessary?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: I can't find a place that uses this default value.

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: That's because there are few places that create v2 relations so far, but when SQL statements and other paths that don't allow you to supply your own schema are added, I think this will be more common. It's okay to remove it, but I don't see much value in the change and I like to keep non-functional changes to a minimum.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: yea I agree we should minimum the non-functional changes, but removing dead code is also good to do. This is really a small change, if we do need the default value in the future, it's very easy to add it back.

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: Either way, it's up to you.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 4]
```diff
@@ -106,7 +106,7 @@ case class StreamingDataSourceV2Relation(
 
 object DataSourceV2Relation {
   private implicit class SourceHelpers(source: DataSourceV2) {
-    def asReadSupport: ReadSupport = {
+    private def asReadSupport: ReadSupport = {
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: Minor: these are effectively private because `SourceHelpers` is private. If we were to move `SourceHelpers` or make it public to some other part of Spark, we would have to revert these changes. So I think it is best to rely on the visibility of `SourceHelpers` instead of making these private.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: makes sense

**Comment by rdblue** [comment]
+1 (non-binding) assuming that tests pass.

**Comment by cloud-fan** [comment]
thanks, merging to master!
